### PR TITLE
Various fixes for preceding-sibling and following-sibling axis

### PR DESCRIPTION
**Pull request by adamretter**
Closes https://github.com/eXist-db/exist/issues/1691

1. This includes many fixes for `preceding-sibling` and `following-sibling` axis in eXist-db. Particularly at the document level.
2. It also offers an optimisation for `following-sibling` where scanning can commence from the current node; Previously scanning would start from the parent node, and traverse all nodes before the current node.
3. It fixes a number of issues with node selection at the document level.
4. Fixes an issue with index data loss for nodes at the document level when reindexing documents.
5. XML:DB Remote API when calling `#getContentAsDOM()` will now return an element node where appropriate, and not always a document node.
6. `DocumentImpl#getNode(NodeId)` now correctly returns all node types, not just elements.

### CODE REVIEW

File: src/org/exist/xmldb/RemoteXMLResource.java [hunk 1]
```diff
@@ -96,12 +120,14 @@ public RemoteXMLResource(
             final int handle,
             final int pos,
             final XmldbURI docId,
-            final Optional<String> id)
+            final Optional<String> id,
+            final Optional<String> type)
```
Comment (dizzzz) [inline src/org/exist/xmldb/RemoteXMLResource.java]: shouldn't Optional's in parameters of constuctors/methods be avoided?

Comment (adamretter) [reply src/org/exist/xmldb/RemoteXMLResource.java]: I have seen some advice that says that yes, and some that doesn't. I am not 100% convinced yet...

File: extensions/indexes/lucene/src/org/exist/xquery/modules/lucene/Query.java [hunk 2]
```diff
@@ -316,13 +318,21 @@ protected Properties parseOptions(Sequence contextSequence, Item contextItem) th
         Sequence optSeq = getArgument(2).eval(contextSequence, contextItem);
         NodeValue optRoot = (NodeValue) optSeq.itemAt(0);
         try {
-            XMLStreamReader reader = context.getXMLStreamReader(optRoot);
+            final int thisLevel = optRoot.getNodeId().getTreeLevel();
+            final XMLStreamReader reader = context.getXMLStreamReader(optRoot);
+            reader.next();
```
[the remark this answers is not in the document]
Comment (dizzzz) [reply extensions/indexes/lucene/src/org/exist/xquery/modules/lucene/Query.java]: @adamretter ?

**Comment by duncdrum** [comment]
Ehm win builds are failing?

**Comment by dizzzz** [comment]
Ow?

**Pull request merged**